ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Monitor.hs

```diff
@@ -63,27 +64,53 @@ governor_BOOTSTRAP_PEERS_TIMEOUT = 15 * 60
 -- state) then, until the node reaches a clean state, this monitoring action
 -- will be disabled and thus churning will be disabled as well.
 --
+-- On the other hand, if Genesis mode is on for the node, this action responds
+-- to changes in ledger state judgement monitoring actions to change the static
+-- set of target peers.
 targetPeers :: (MonadSTM m, Ord peeraddr)
             => PeerSelectionActions peeraddr peerconn m
             -> PeerSelectionState peeraddr peerconn
             -> Guarded (STM m) (TimedDecision m peeraddr peerconn)
-targetPeers PeerSelectionActions{readPeerSelectionTargets}
+targetPeers PeerSelectionActions{ readPeerSelectionTargets,
+                                  peerTargets = ConsensusModePeerTargets {
+                                    praosTargets,
+                                    genesisSyncTargets } }
             st@PeerSelectionState{
               publicRootPeers,
               localRootPeers,
               targets,
-              ledgerStateJudgement,
               bootstrapPeersFlag,
-              hasOnlyBootstrapPeers
+              hasOnlyBootstrapPeers,
+              ledgerStateJudgement,
+              consensusMode
             } =
     Guarded Nothing $ do
-      targets' <- readPeerSelectionTargets
-      check ( isNodeAbleToMakeProgress bootstrapPeersFlag
-                                       ledgerStateJudgement
-                                       hasOnlyBootstrapPeers
-            && targets' /= targets
-            && sanePeerSelectionTargets targets'
-            )
+      churnTargets <- readPeerSelectionTargets
+      -- Genesis consensus mode:
+      -- we check if targets proposed by churn are stale
+      -- in the sense that they are the targets for
+      -- opposite value of the current ledger state judgement.
+      -- This indicates that we aren't churning currently, and
+      -- furthermore it means that the ledger state has flipped since
+      -- we last churned. Therefore we can't set the targets from
+      -- the TVar, and instead we set the appropriate targets
+      -- for the mode we are in.
+      let targets' =
+            case (ledgerStateJudgement, consensusMode) of
+              (YoungEnough, GenesisMode)
+                | churnTargets == genesisSyncTargets ->
+                  praosTargets
+              (TooOld, GenesisMode)
+                | churnTargets == praosTargets ->
+                  genesisSyncTargets
+              _otherwise -> churnTargets
```

Comment:
That's clever.  We just need to make sure it's consistent with churn.

Even if churn is ongoing and targets are set to either `genesisSyncTargets` or `praosTargets` we will update the targets.  But that's fine as long as churn is not caching anything when it starts it's loop, which I think it is the case.  Is this right?

Reply:
Correct, before churn modifies the targets TVar it checks FetchMode which is synchronized with LedgerStateJudgement by @bolt12's patch. So even while it's going it can switch to churning the right target basis.